docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,247 @@
+* MeshAccessLog
+* MeshCircuitBreaker
+* MeshFaultInjection
+* MeshHealthCheck
+* MeshHttpRoute
+* MeshProxyPatch
+* MeshRateLimit
+* MeshRetry
+* MeshTimeout
+* MeshTrace
+* MeshTrafficPermission
+
+A client sends DiscoveryRequest to the server and waits for the change. The server stores the state in a state-of-the-world cache and sends the response whenever there is a change. It works well in the case of Global to Zone synchronization
+but might have scalability issues for synchronization Zone to Global with large deployments. Zone control-plane needs to send all the Dataplane resources to the global whenever there is a change. That might be a bottleneck that can happen in large deployments. We can use [`Incremental xDS`](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol) which is Envoy's implementation of cache that instead of sending all the resources sends only changes. That could reduce the bandwidth usage.
+
+```proto
+service GlobalKDSService {
+ ...
+ rpc ZoneToGlobalSync(stream envoy.service.discovery.v3.DeltaDiscoveryRequest)
+     returns (stream envoy.service.discovery.v3.DeltaDiscoveryResponse);
+
+ rpc GlobalToZoneSync(stream envoy.service.discovery.v3.DeltaDiscoveryRequest)
+     returns (stream envoy.service.discovery.v3.DeltaDiscoveryRequest);
+}
+
+```
+
+In the first iteration, we can introduce only `ZoneToGlobalSync` with delta xDS, and in the future, we could introduce `GlobalToZoneSync` and also use delta xDS. In the code, I've noticed that we are using our copy of the Snapshot cache and the code is a bit outdated. As a part of that task, we should check if we can change to use `go-control-plane` cache implementation or copy and update our current implementation. 
+
+Maybe not in the first step but we could consider if we can benefit from using different cache implementation `LinearCache` or `MuxCache` for snapshot generation.
+
+### Prioritize operational informations
+
+Certain resources require more frequent synchronization than others. For example, Policies, ZoneIngress, and ZoneEgress must be updated immediately, whereas Insights and Dataplanes objects may be delayed and can be aggregated. To address this, we could consider implementing an option to specify the interval at which the snapshot for specific resources is recalculated. To avoid recalculating changes for all synchronized resources, we need a cache per resource type, which `SimpleCache` should allows us to achieve. In addition, we could set up an RPC service to monitor changes for all resources on separate streams, allowing changes to be processed independently.
+
+### Change the namespace of synchronized resources from `default` to `kuma-system`
+
+Resources synchronized from Zone to Global are created by default in a `default` namespace and we should change it in the new version. This is a breaking change because upgrade won't be able to find resources and needs to duplicate them from zone and store them in the `kuma-system` namespace.
```

Comment:
I guess it would just get them from zone right? But the zone will be updated later.

Do we have an idea how we could make this graceful? (Read from the kds zone connection which version is used and read from the right namespace?)
Also how do we delete old dataplanes after upgrade?

Reply:
In the code it looks like:
1. Upgrade global control-plane
2. Global control-plane requests for Dataplanes
3. Dataplanes are sent
4. Global control-plane check if the store is k8s and it is store them in the specific namespace

I feel like Global (on the kds level) doesn't need to request for resources in a specific namespace < https://github.com/kumahq/kuma/blob/master/pkg/kds/global/components.go#L124

About delete: I feel like we can provide some upgrade note saying: `kubectl delete Dataplanes -n default`

Reply:
Yes upgrade notes makes a lot of sense. It's imperfect but probably good enough and avoids too much logic in the CP for a one of upgrade.

I guess for the duplicate between `default` and system namespace of DPP/Insights we could always do the deduping in `pkg/insights/resyncer.go` no? That way we contain the complexity a little though dataplanes would still show twice in the api (so in the GUI too).

Another option is to do the deduping in the pkg/plugins/resources/k8s/store.go List directly for these specific types?


OR when a zone connects with the new KDS it first connects once with the old api and sends an empty SOTW?

Reply:
> Another option is to do the deduping in the pkg/plugins/resources/k8s/store.go List directly for these specific types?

I feel like listing data from the specific namespace might be the best option to do, we can avoid some custom logic, and just get the correct data

Reply:
> This is a breaking change because upgrade won't be able to find resources and needs to duplicate them from zone and store them in the `kuma-system` namespace.

Why is this a breaking change exactly? What breaks? It's not clear to me.

Do you mean the CP won't be able to find resources? Will then an empty resource set be synced to other zone CPs?

Reply:
@jakubdyszkiewicz explained to me that not all resources are namespace scoped and some are not stored in the namespace. We cannot change it that easily. One of the breaking changes might be `ZoneIngress` resource that might be in any namespace, also it is synced from zone to global and global to zone. In case we have an old entry we might sync to the zone's old resource which can point to old/not working zone ingress and break the traffic.

Reply:
https://github.com/kumahq/kuma/pull/6102#discussion_r1122218493